feat: add method to deallocate reactNativeFactory instance for iOS

californiandreamer opened the pull request:
### Summary

* Adds public iOS API stopReactNative() to deallocate the React Native runtime created by RCTReactNativeFactory and allow clean re‑initialization later. Fixes [#134](https://github.com/callstack/react-native-brownfield/issues/134).
* Old Architecture: calls [bridge.invalidate()](https://reactnative.dev/docs/legacy/native-modules-ios#invalidate) to tear down JS and native modules.
* New Architecture: releases the runtime by dropping strong references (no public stop API on `RCTHost`).
* Prevents crashes when pushing a React Native screen after stopping.
* Removes stale lazy caching of `rootViewFactory` to avoid retaining the old runtime.
* Documentation updates for the new API and usage:
   * docs/SWIFT.md
   * docs/OBJECTIVE_C.md
* Backwards compatible: existing `startReactNative()` API remains; `view()` now self‑heals if RN was previously stopped.

### Test plan

* Press the “Stop React Native” (to check app won't crash if `stopReactNative` called before RN was initialize)
* Push RN screen → pop → tap the “Stop React Native” button → push again (no crash).
* Repeat with both architectures toggled.
* Watch the Xcode Instruments graph to confirm RN objects are released after stop (RSS may not drop immediately).

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds an iOS `stopReactNative()` API to tear down the RN runtime, simplifies factory creation/usage, and updates docs and the Swift example.
> 
> - **iOS SDK (`ios/ReactNativeBrownfield.swift`)**:
>   - **New API**: `stopReactNative()` invalidates the bridge, removes observers, and clears the factory; safe to call multiple times.
>   - **Factory lifecycle**: Replaces cached `rootViewFactory` with a lazy `factory` creator; `view(...)` now resolves via `factory`; `startReactNative(..., launchOptions:)` simplified.
> - **Docs**:
>   - Add `stopReactNative` reference, description, and examples in `docs/SWIFT.md` and `docs/OBJECTIVE_C.md`.
> - **Example app** (`example/swift/App.swift`):
>   - Add "Stop React Native" button calling `ReactNativeBrownfield.shared.stopReactNative()`; minor UI tweaks.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit aab133981ccb4ab8b8d7e16ef8f55c68ae766049. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

californiandreamer committed:
feat: add stopReactNative usage example

okwasniewski commented on ios/ReactNativeBrownfield.swift:
```diff
@@ -136,6 +143,25 @@ class ReactNativeBrownfieldDelegate: RCTDefaultReactNativeFactoryDelegate {
+      return
+    }
+
+    guard let factory = reactNativeFactory else { return }
```
```suggestion
    guard let reactNativeFactory else { return }
```

okwasniewski commented on ios/ReactNativeBrownfield.swift:
```diff
@@ -32,6 +32,13 @@ class ReactNativeBrownfieldDelegate: RCTDefaultReactNativeFactoryDelegate {
   private var onBundleLoaded: (() -> Void)?
   private var delegate = ReactNativeBrownfieldDelegate()
   
+  private func checkFactoryInitialized(launchOptions: [AnyHashable: Any]? = nil) {
+    if reactNativeFactory == nil {
+      delegate.dependencyProvider = RCTAppDependencyProvider()
+      self.reactNativeFactory = RCTReactNativeFactory(delegate: delegate)
+    }
+  }
```
Can we create w getter on the factory that initizalizes it if it's nil? It should remove the need to call this function

okwasniewski commented on ios/ReactNativeBrownfield.swift:
```diff
@@ -136,6 +143,25 @@ class ReactNativeBrownfieldDelegate: RCTDefaultReactNativeFactoryDelegate {
+
+    guard let factory = reactNativeFactory else { return }
+
+    factory.bridge?.invalidate()
+
+    NotificationCenter.default.removeObserver(self)
+    onBundleLoaded = nil
+
+    reactNativeFactory = nil
```
Did you test the scenario of what happens when react native view is still shown while calling this? Let's make sure it doesn't crash the app.

okwasniewski replied:
Maybe there is a way to detect current react native surfaces in the app and show a warning if the user tries to do this with views still on screen

okwasniewski reviewed (commented):
Thanks for working on this! 🚀

Overall looks good, left couple of questions.

californiandreamer commented:
@okwasniewski, I refactored the factory a bit and added some additional checks due to Cursor Bugbot comments.

When it comes to the `stopReactNative()` call on the RN screen, the app is not crashing but just goes white screen. We can think about a bit better fallback. I spent some time on that and didn't find a solution yet. Will dig deeper 😄 

Simulator Screen Recording - iPhone 16 Pro - 2025-09-24 at 22.52.53.mp4

okwasniewski commented:
@californiandreamer This looks like a expected behavior if you stop react native, let's ship it.

okwasniewski commented on ios/ReactNativeBrownfield.swift:
```diff
@@ -111,29 +121,54 @@ class ReactNativeBrownfieldDelegate: RCTDefaultReactNativeFactoryDelegate {
+        DispatchQueue.main.async { [weak self] in
+          self?.jsLoaded(Notification(name: notificationName))
+        }
       }
```
This is changing the previous behavior. jsLoaded should be called by React Native sending this notification

okwasniewski commented on ios/ReactNativeBrownfield.swift:
```diff
@@ -111,29 +121,54 @@ class ReactNativeBrownfieldDelegate: RCTDefaultReactNativeFactoryDelegate {
+      return
+    }
+
+    guard let factory = reactNativeFactory else {
```
nit: `guard let reactNativeFactory else`

californiandreamer committed:
feat: add stopReactNative method

californiandreamer committed:
chore: refactor factory initializer

californiandreamer committed:
chore: add storedLaunchOptions handler

hurali97 commented:
Closing in favor of https://github.com/callstack/react-native-brownfield/pull/301
